diff --git a/html/src/components/terminal/xterm/index.ts b/html/src/components/terminal/xterm/index.ts
index a163ef6..9e7dac3 100644
--- a/html/src/components/terminal/xterm/index.ts
+++ b/html/src/components/terminal/xterm/index.ts
@@ -463,6 +463,12 @@ export class Xterm {
                         if (enabled) console.log('[ttyd-mod] trim selection enabled');
                     }
                     break;
+                case 'enablePasteOnCtrlV':
+                    if (value) {
+                        const enabled = this.enablePasteOnCtrlV();
+                        if (enabled) console.log('[ttyd-mod] paste on Ctrl+V enabled');
+                    }
+                    break;
                 default:
                     console.log(`[ttyd] option: ${key}=${JSON.stringify(value)}`);
                     if (terminal.options[key] instanceof Object) {
@@ -567,4 +573,70 @@ export class Xterm {
             console.warn('[ttyd-mod] Opening link blocked as opener could not be cleared');
         }
     }
+
+    @bind
+    private enablePasteOnCtrlV(): boolean {
+        const { terminal } = this;
+
+        const isCtrlV = (e: KeyboardEvent): boolean =>
+            e.ctrlKey &&
+            !e.shiftKey &&
+            !e.altKey &&
+            !e.metaKey &&
+            (e.code === 'KeyV' || e.key.toLowerCase() === 'v') &&
+            e.type === 'keydown' &&
+            !e.repeat &&
+            !e.isComposing;
+
+        terminal.attachCustomKeyEventHandler((e: KeyboardEvent): boolean => {
+            // returns whether the event should be processed by xterm.js
+            if (!isCtrlV(e)) return true;
+
+            const ua = navigator.userAgent;
+            if (ua.includes('Firefox')) return firefox(e);
+            if (ua.includes('Chrome')) return chromium(e);
+            if (ua.includes('Chromium')) return chromium(e);
+
+            console.log('[ttyd-mod] this browser is not supported: ' + ua);
+            return true;
+        });
+        return true;
+
+        function firefox(e: KeyboardEvent): boolean {
+            void e;
+            // In the case of Firefox, if false is returned here when
+            // Ctrl+V is pressed, Firefox would paste by itself. Other
+            // conditions also required, such as e.preventDefault() not
+            // being called inside the keydown event handlers.
+            return false;
+        }
+
+        function chromium(e: KeyboardEvent): boolean {
+            withTimeout(navigator.clipboard.readText())
+                .then((text: string) => {
+                    if (text) terminal.paste(text);
+                })
+                .catch(err => {
+                    console.error('[ttyd-mod] readText():', err);
+                });
+            e.preventDefault();
+            return false;
+        }
+
+        // Timeout Helper
+        function withTimeout<T>(p: Promise<T>, ms = 200): Promise<T> {
+            return new Promise<T>((resolve, reject) => {
+                const id = setTimeout(() => {
+                    reject(new Error('timeout'));
+                }, ms);
+                p.then(v => {
+                    clearTimeout(id);
+                    resolve(v);
+                }).catch(e => {
+                    clearTimeout(id);
+                    reject(e);
+                });
+            });
+        }
+    }
 }
